<!--
  Copyright 2016 The Cartographer Authors

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
<launch>
  <arg name="namespace"               default="$(env UAV_NAMESPACE)"/>
  
  <!-- Simulation parameters -->
  <arg name="simulation"              default="false"/>
  <arg name="slam_ns"                 default="map"/>

  <!-- Cartographer topic names -->
  <arg name="imu"                     default="mavros/imu/data"/>
  <arg name="fix"                     default="mavros/global_position/global"/>
  <arg name="points2"                 default="velodyne_points"/>
  <arg name="transform"               default="/pozyx/measured"/>
  
  <!-- Cartographer configuration parameters -->
  <arg name="configuration_directory" default="$(find uav_ros_general)/cartographer_config"/>
  <arg name="configuration_basename"  default ="second_tuning_new_params.lua"/>

  <!-- Starter parameters -->
  <arg name="enable_starter"          default="false"/>
  <arg name="starter_odom"            default="mavros/global_position/local"/>

  <group ns="$(arg namespace)">

    <!-- Launch cartographer_ros node -->
    <node 
      name="cartographer_node" pkg="cartographer_ros"
      type="cartographer_node" args="
          -start_trajectory_with_default_topics=false
          -configuration_directory $(arg configuration_directory)
          -configuration_basename $(arg configuration_basename)"
      output="screen">

        <remap from="imu"       to="$(arg imu)"/>
        <remap from="points2"   to="$(arg points2)"/>
        <remap from="fix"       to="$(arg fix)"/>
        <remap from="transform" to="$(arg transform)"/>
    </node>

    <!-- Launch slam starter -->
    <node 
      name="SLAM_starter" pkg="cartographer_ros" type="slam_starter.py"
      args="$(arg configuration_directory) $(arg configuration_basename)">
        
        <remap from="odom" to="$(arg starter_odom)"/>
    </node>

    <!-- Publishes uav/cartographer/pose with the different namespace for sim and online -->
    <group if="$(arg simulation)">
      <node name="pose_from_tf" pkg="uav_ros_general" type="pose_from_tf.py" output="screen">
        <param name="map_frame"       value="$(arg namespace)/$(arg slam_ns)/map"/>
        <param name="tracking_frame"  value="$(arg namespace)/$(arg slam_ns)/base_link"/>
      </node>
    </group>

    <group unless="$(arg simulation)">
      <node name="pose_from_tf" pkg="uav_ros_general" type="pose_from_tf.py" output="screen">
        <param name="map_frame"       value="$(arg namespace)/map"/>
        <param name="tracking_frame"  value="$(arg namespace)/base_link"/>
      </node>
    </group>

  </group>
</launch>

